services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/AttributeName.java

```diff
@@ -0,0 +1,140 @@
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * High-level representation of a DynamoDB 'attribute name' that can be used in various situations where the API requires
+ * or accepts an Attrbute Name. In addition various convenience methods are provided to help manipulate expressions.
+ * <p>
+ * At a minimum, an AttributeName must contain a string that represents the name of the Attribute.
+ * <p>
+ * Optionally, nested attribute names can be added with tokens using the addNestedAttributeName .
+ * Every call to addNestedAttributeName represent next level of Nesting Attribute.
+ * If we want to denote three level of Nesting attribute name then addNestedAttributeName should be called thrice.
+ *
+ * <p>
+ * Example, if we want to represent a two level nested attribute outerAttribute.innerAttributeLevel1.innerAttributeLevel2
+ * then it can be represented as :-
+ * {@code
+ * AttributeName myAttributeName = AttributeName.builder()
+ * .name("outerAttribute).addNested("innerAttributeLevel1).nestedAttributeName("innerAttributeLevel2").build();
+ * }
+ */
+@SdkPublicApi
+public final class AttributeName {
+
+    private final String name;
+    private final List<String> nestedAttributeNames;
+
+    private AttributeName(String name, List<String> nestedAttributeNames) {
+        this.name = name;
+        this.nestedAttributeNames = nestedAttributeNames != null
+                ? Collections.unmodifiableList(nestedAttributeNames) : null;
+    }
+
+    /**
+     * Construct a new AttributeNameBuilder.
+     *
+     * @return a new AttributeNameBuilder.
+     */
+    public static AttributeNameBuilder builder() {
+        return new Builder();
+    }
+
+    /**
+     * @return Name of the attribute.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * @return List of nested attributes , each entry in the list represent one level of nesting.
+     */
+    public List getNestedAttributeNames() {
+        return nestedAttributeNames;
+    }
+
+    /**
+     * First Builder Step in charge of setting Attribute Name.
+     * Next Builder available : addNestedAttribute.
+     */
+    public interface AttributeNameBuilder {
+        BuildStep name(String attributeToProject);
+    }
+
+    /**
+     * This is the final step in charge of building the AttributeName Object.
+     */
+    public interface BuildStep {
+        BuildStep addNested(String nestedAttributeName);
+
+        AttributeName build();
+    }
```

Comment:
We avoid this builder pattern, we think it creates more complexity and confusion than it solves. We should try and design a simple builder that intuitively works for both use cases. This may not be the best solution but whatever solution we come up with has to work for all the scenarios. E.g.:

```
// All of these are equivalent
AttributeName.builder().name("foo").addNestedNames("bar").build();
AttributeName.builder().addNestedNames("foo", bar");
AttributeName.builder().addNestedNames("foo").addNestedNames("bar");
```